enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,254 @@
+
+## Design Details
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name? 
+- In which namespace the new operator components should be hosted?
+
+### current implmentation 
+In current implementation, the [networking infra componenets](https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md) run as static pods through the MCO.
+
+### Suggested design
+
+To support early clustering requirements Keepalived will continue running as static pods through MCO, additionally, a new dnsmasq service (also deployed through MCO) will run in each node while HAProxy, CoreDNS-MDNS and MDNS-publisher will run as daemonset pods through the new operator.
+
+The following [diagram](https://docs.google.com/drawings/d/1a8K16vcaBB6SIF0bN8DaTj4ltjBxVsVNjJN5NCd-U0A/edit?usp=sharing) reflects the suggested changes.
+
+#### Self hosted DNS services
+
+- The self hosted DNS functionality should be composed of the following components in each node:
+    - A NetworkManager dispatcher script used to configure resolv.conf to point at the node's public IP address, see [self hosted DNS design](https://github.com/openshift/installer/blob/f6bb83784a835e1f4955ef41916793c0948fd4a1/docs/design/baremetal/networking-infrastructure.md#internal-dns) for more details.
+    - Systemd service running dnsmasq
+    - CoreDNS-mDNS pod
+- The CHNSO will be responsible only for the deployment of the CoreDNS pods while the NetworkManager dispatcher script and dnsmasq Systemd service will run through MCO.
+- To avoid port conflict, Dnsmasq service should listen on DNS well-known port (53) and CoreDNS-MDNS should listen on other port (e.g: 5353). 
+- The new Dnsmasq service should be responsible to resolve 'api-int' (to API-VIP value provided in install-config) hostname when CoreDNS-MDNS pod isn't running on the node.
+- CoreDNS-MDNS should register himself as an upstream server in dnsmasq [via dbus interface](https://github.com/imp/dnsmasq/blob/master/dbus/DBus-interface#L51) for ClusterName.DomainName
+ 
+#### Self hosted API Loadbalancer
+
+- The self hosted API loadbalancer will run similarly to the current mode with one difference: HAProxy pods will run by CHSNO instead of the static pods.
+- API should be available even when HAProxy pods don't run.
+- Additional functionality aspects of the self-hosted Loadbalancer maybe further analyzed upon need in a separate enhancement. 
+
+
+#### Disaster recovery support
+
+In order to support disaster recovery kubelet, kube-controller-manager and kube-scheduler should be able to communicate with kube-api even when all CHNSO pods are shutdown.
+
+The following steps will ensure that kube-api is available post disaster recovery:
+- Dnsmasq service and Keepalived static pods will start running on master nodes
+- API-VIP address will be assigned to one of the master nodes by Keepalived
+- Dnsmasq should resolve api-int hostname to API-VIP address
```

Comment:
I don't see how this helps. Today the API traffic goes through the haproxy https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md#api-load-balancing. And there are technical reasons (graceful termination among others) that we need that and cannot switch to a VIP directly. Via DNS you cannot map ports, only names<->IPs.

Reply:
The status today (installer docs should be updated):
- API goes first to API-VIP
- If HAProxy is running  API redirected to LB. 

We keep the same logic as today, just moving HAProxy to run as daemonset pod instead of static pod.